src/providers/user-data/user-data.ts

```diff
@@ -207,6 +207,21 @@ export class UserDataProvider {
     return this.saveProfiles();
   }
 
+  public setWalletLabel(wallet: Wallet, label: string): boolean {
+    if (!wallet || wallet.label === label) {
+      return;
+    }
+
+    const exists = lodash.some(this.currentProfile.wallets, w => label && w.label && w.label.toLowerCase() === label.toLowerCase());
```

Comment:
This could be done separately. Just to not confuse the function name, so we will have something like `hasWalletLabel` and `setWalletLabel`.

Reply:
but this checks not only if it has "any" wallet label, but instead if it has the _same_ wallet label.
Do you still want to extract a function? If so: Did you mean to call from outside first "WalletLabelExists?" and then only if it doesn't we call `SetWalletLabel` (where we don't check if a wallet with the same label exists anymore)? If so I find it a little strange and it's imho safer the way it's now.

Reply:
I'm just thinking it's strange to return 'undefined', a boolean (true) or an observable in the same function.

Reply:
I see  what you mean now, however an `Observable` is never returned, but 2 `return false` are missing.

So what do you prefer of these solutions:
1. Just leave the method as it is and `return false`
```
  public setWalletLabel(wallet: Wallet, label: string): boolean {
    if (!wallet || wallet.label === label) {
      return false;
    }

    const exists = lodash.some(this.currentProfile.wallets, w => label && w.label && w.label.toLowerCase() === label.toLowerCase());

    if (exists) {
      return true;
    }

    wallet.label = label;
    this.saveWallet(wallet);
    return false;
  }
```

2. Probably what you meant:

```
  public setWalletLabel(wallet: Wallet, label: string): void {
    if (!wallet || wallet.label === label) {
      return;
    }

    wallet.label = label;
    this.saveWallet(wallet);
  }

  public doesWalletWithLabelExist(wallet: Wallet, label: string): boolean {
    if (!label || !wallet || wallet.label === label) {
      return false;
    }

    return lodash.some(this.currentProfile.wallets, w => w.label && w.label.toLowerCase() === label.toLowerCase());
  }

//.... and the call in the dashboard class:
      if (this.userDataProvider.doesWalletWithLabelExist(this.wallet, label)) {
        this.toastProvider.error({key: 'WALLETS_PAGE.LABEL_EXISTS', parameters: {label: label}} as TranslatableObject, 3000);
      } else {
        this.userDataProvider.setWalletLabel(this.wallet, label);
      }
```

Reply:
Return `true` makes more sense to me:

```
public setWalletLabel(wallet: Wallet, label: string): boolean {
    if (!wallet || wallet.label === label) {
      return;
    }

    const exists = lodash.some(this.currentProfile.wallets, w => label && w.label && w.label.toLowerCase() === label.toLowerCase());

    if (exists) {
      return false;
    }

    wallet.label = label;
    this.saveWallet(wallet);
    return true;
  }
```

Reply:
Yes it would. But then if you don't specify a wallet or the label is the same as before, it would also return false (or undefined in your code), because of this lines:
```
    if (!wallet || wallet.label === label) {
      return;
    }
```
And this would then mean that the error message `Duplicate label` would be displayed.

Maybe this whole method is ugly and should `throw` an `exception` (`IDuplicateLabelError`) instead of returning booleans at all?

Reply:
Does not that solve it?
```
const labelExists = this.userDataProvider.setWalletLabel(this.wallet, label);

if (labelExists === false) {
```

Reply:
You mean an explicit `false` check (so `undefined` doesn't count)?
I guess this would - and if you're happy with that, I can also gladly agree to this :)

Reply:
Yes, it's fine for me.